qt/widgets/instrumentview/src/Shape2D.cpp

```diff
@@ -620,6 +623,497 @@ std::string Shape2DRing::saveToProject() const {
+      m_outerRadius = 1.01 * m_innerRadius;
+    }
+    break;
+  case 1:
+    m_innerRadius = distanceBetween(to_center, QPointF(0, 0));
+    if (m_outerRadius < m_innerRadius) {
+      m_innerRadius = 0.99 * m_outerRadius;
+    }
+    break;
+  case 2:
+    newAngle = atan2(to_center.y(), to_center.x());
+    if (newAngle < 0)
+      newAngle += 2 * M_PI;
+
+    // conditions to prevent the startAngle from going over the endAngle
+    if ((m_startAngle < m_endAngle && newAngle >= m_endAngle &&
+         std::abs(newAngle - m_startAngle) < 1) ||
+        (newAngle < m_endAngle && m_startAngle < m_endAngle &&
+         std::abs(newAngle - m_startAngle) > 5.28 && newAngle < m_startAngle) ||
+        (newAngle > m_endAngle && m_startAngle > m_endAngle &&
+         std::abs(newAngle - m_startAngle) > 5.28 && newAngle < m_startAngle)) {
+
+      newAngle = m_endAngle - 0.00001;
+
+      if (newAngle < 0) {
+        newAngle += 2 * M_PI;
+      }
+
+    } else if ((m_startAngle > m_endAngle && newAngle <= m_endAngle &&
+                std::abs(newAngle - m_startAngle) < 1) ||
+               (newAngle > m_endAngle && m_startAngle > m_endAngle &&
+                std::abs(newAngle - m_startAngle) > 5.28 &&
+                newAngle > m_startAngle) ||
+               (newAngle < m_endAngle && m_startAngle < m_endAngle &&
+                std::abs(newAngle - m_startAngle) > 5.28 &&
+                newAngle > m_startAngle)) {
+
+      newAngle = m_endAngle + 0.0001;
+
+      if (newAngle >= 2 * M_PI) {
+        newAngle -= 2 * M_PI;
+      }
+    }
+
+    m_startAngle = newAngle;
+    break;
+
+  case 3:
+    newAngle = atan2(to_center.y(), to_center.x());
+    if (newAngle < 0)
+      newAngle += 2 * M_PI;
+
+    if ((m_endAngle < m_startAngle && newAngle >= m_startAngle &&
+         std::abs(newAngle - m_endAngle) < 1) ||
+        (newAngle < m_startAngle && m_endAngle < m_startAngle &&
+         std::abs(newAngle - m_endAngle) > 5.28 && newAngle < m_endAngle) ||
+        (newAngle > m_startAngle && m_endAngle > m_startAngle &&
+         std::abs(newAngle - m_endAngle) > 5.28 && newAngle < m_endAngle)) {
+
+      newAngle = m_startAngle - 0.00001;
+
+      if (newAngle < 0) {
+        newAngle += 2 * M_PI;
+      }
+
+    } else if ((m_endAngle >= m_startAngle && newAngle <= m_startAngle &&
+                std::abs(newAngle - m_endAngle) < 1) ||
+               (newAngle >= m_startAngle && m_endAngle >= m_startAngle &&
+                std::abs(newAngle - m_endAngle) > 5.28 &&
+                newAngle > m_endAngle) ||
+               (newAngle < m_startAngle && m_endAngle < m_startAngle &&
+                std::abs(newAngle - m_endAngle) > 5.28 &&
+                newAngle > m_endAngle)) {
+
+      newAngle = m_startAngle + 0.0001;
+      if (newAngle >= 2 * M_PI) {
+        newAngle -= 2 * M_PI;
+      }
+    }
+
+    m_endAngle = newAngle;
+    break;
+  }
+  resetBoundingRect();
+}
+
+QStringList Shape2DSector::getDoubleNames() const {
```

Comment:
What about the center?

Reply:
There is now control point in the center, so it is not relevant here.